### PR TITLE
Add Twitter OAuth login module(login with twitter)

### PR DESCRIPTION
**Pull request by Kumo-YZX**
##### Description

1. Add TwitterOAuth method to python-social-auth module by edit AUTHENTICATION_BACKENDS. And add TWITTER_CONSUMER_KEY, TWITTER_CONSUMER_SECRET, TWITTER_ACCESS_TOKEN, TWITTER_ACCESS_SECRET to settings.py so that we can use OAuth module.

2. Add twitter-login url to app/urls.py and set its dispatcher to twitter_login() method in dashboard/views. 
And add twitter_login() function to dashboard/views.

3. Add self-defined social_details() function with the mechanism to add twitter prefix to the social-auth pipeline.

4. Add a twitter module to get Twitter user info from Twitter API.

5. Distinguish get_user() function in Git and Twitter module by setting the alias as get_github_user() and get_twitter_user().

6. Expand the avatar module to handle both GitHub and Twitter avatars. Allow users to set avatar using 
their Twitter avatar.

7. Expand save_profile() and sync_profile() function to add Twitter support.

8. Add "Login with Twitter" button to the nav bar. Add "twitter" text to pages that related to Twitter function.
##### Refers/Fixes

Fixes: #4111

##### Testing

I have tested the code with my twitter app. The login and avatar module works well. The only test left is the kudos module because there is something wrong with my ipfs network. But kudos may work if the user is logged in?

### CODE REVIEW

File: app/retail/templates/home/nav.html [hunk 1]
```diff
@@ -100,12 +100,16 @@
       {% if user.is_authenticated %}
         {% include 'shared/nav_auth.html' with source='authed' %}
       {% else %}
+        <a href="{% url 'social:begin' 'twitter' %}?next=/" class="nav-link interior">
+          <i class="fab fa-github mr-1" role="aria-hidden"></i>
```
Comment (owocki) [inline app/retail/templates/home/nav.html]: i dont think we need this login link in the nav.  thanks tho!

File: app/app/settings.py [hunk 2]
```diff
@@ -532,22 +534,37 @@
...
+# Twitter Integration
+TWITTER_CONSUMER_KEY = env('TWITTER_CONSUMER_KEY', default='')  # TODO
+TWITTER_CONSUMER_SECRET = env('TWITTER_CONSUMER_SECRET', default='')  # TODO
+TWITTER_ACCESS_TOKEN = env('TWITTER_ACCESS_TOKEN', default='')  # TODO
+TWITTER_ACCESS_SECRET = env('TWITTER_ACCESS_SECRET', default='')  # TODO
+TWITTER_USERNAME = env('TWITTER_USERNAME', default='')  # TODO
+
 # Social Auth
 LOGIN_URL = 'gh_login'
+LOGIN_URL = 'twitter_login'
```
Comment (owocki) [inline app/app/settings.py]: can i still login with github too?

File: app/app/settings.py [hunk 3]
```diff
@@ -532,22 +534,37 @@
 GITHUB_API_TOKEN = env('GITHUB_API_TOKEN', default='')  # TODO
 GITHUB_APP_NAME = env('GITHUB_APP_NAME', default='gitcoin-local')
 
+# Twitter Integration
```
Comment (owocki) [inline app/app/settings.py]: when i try to login i get the following exception
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/django/contrib/staticfiles/handlers.py", line 65, in __call__
    return self.application(environ, start_response)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/wsgi.py", line 142, in __call__
    response = self.get_response(request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 78, in get_response
    response = self._middleware_chain(request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 36, in inner
    response = response_for_exception(request, exc)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 90, in response_for_exception
    response = handle_uncaught_exception(request, get_resolver(get_urlconf()), sys.exc_info())
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 125, in handle_uncaught_exception
    return debug.technical_500_response(request, *exc_info)
  File "/usr/local/lib/python3.7/site-packages/django_extensions/management/technical_response.py", line 37, in null_technical_500_response
    six.reraise(exc_type, exc_value, tb)
  File "/usr/local/lib/python3.7/site-packages/six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 126, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/usr/local/lib/python3.7/site-packages/django/core/handlers/base.py", line 124, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/usr/local/lib/python3.7/site-packages/django/views/decorators/cache.py", line 44, in _wrapped_view_func
    response = view_func(request, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/social_django/utils.py", line 49, in wrapper
    return func(request, backend, *args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/social_django/views.py", line 23, in auth
    return do_auth(request.backend, redirect_name=REDIRECT_FIELD_NAME)
  File "/usr/local/lib/python3.7/site-packages/social_core/actions.py", line 27, in do_auth
    return backend.start()
  File "/usr/local/lib/python3.7/site-packages/social_core/backends/base.py", line 35, in start
    return self.strategy.redirect(self.auth_url())
  File "/usr/local/lib/python3.7/site-packages/social_core/backends/oauth.py", line 167, in auth_url
    token = self.set_unauthorized_token()
  File "/usr/local/lib/python3.7/site-packages/social_core/backends/oauth.py", line 223, in set_unauthorized_token
    token = self.unauthorized_token()
  File "/usr/local/lib/python3.7/site-packages/social_core/backends/oauth.py", line 247, in unauthorized_token
    method=self.REQUEST_TOKEN_METHOD
  File "/usr/local/lib/python3.7/site-packages/social_core/backends/base.py", line 234, in request
    response.raise_for_status()
  File "/usr/local/lib/python3.7/site-packages/requests/models.py", line 940, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Authorization Required for url: https://api.twitter.com/oauth/request_token

```

File: app/retail/views.py [hunk 4]
```diff
@@ -16,13 +16,17 @@
     along with this program. If not, see <http://www.gnu.org/licenses/>.
 
 '''
+
+
 import logging
 from json import loads as json_parse
+from json import load as json_load
+from json import dump as json_dump
```
Comment (owocki) [inline app/retail/views.py]: none of these seem used?

File: app/retail/templates/home/nav.html (hunk 1 above)
Comment (Kumo-YZX) [reply app/retail/templates/home/nav.html]: Oh I see.

File: app/retail/views.py (hunk 4 above)
Comment (Kumo-YZX) [reply app/retail/views.py]: Oh, It is used by previous version and is useless now. Just delete it.

File: app/app/settings.py (hunk 2 above)
Comment (Kumo-YZX) [reply app/app/settings.py]: I found it working even with these lines commented. Maybe it is useless?
image

File: app/app/settings.py (hunk 3 above)
Comment (Kumo-YZX) [reply app/app/settings.py]: I have got these error before. Check the settings in .env file and reload the program?
image

Comment (Kumo-YZX) [reply app/app/settings.py]: Sometimes Twitter API throw 403 or 401 error when the keys are just set. Maybe we need to wait for a while and reload.

File: app/retail/templates/home/nav.html (hunk 1 above)
Comment (owocki) [reply app/retail/templates/home/nav.html]: can u pls remove

**Comment by owocki** [comment]
hm... i wonder if we need to add any changes to the code that links users to profiles and visa versa.. a lot of the code all over the app assumes that a user has a link to a `dashboard.profile` in it... i worry that that code will break now.

maybe to solve this we cordon off most of the app to twitter-logins, and only allow people to access the airdrop functionality from twitter

File: app/retail/templates/home/nav.html (hunk 1 above)
Comment (Kumo-YZX) [reply app/retail/templates/home/nav.html]: already updated.

**Comment by Kumo-YZX** [comment]
> hm... i wonder if we need to add any changes to the code that links users to profiles and visa versa.. a lot of the code all over the app assumes that a user has a link to a `dashboard.profile` in it... i worry that that code will break now.
> 
> maybe to solve this we cordon off most of the app to twitter-logins, and only allow people to access the airdrop functionality from twitter

Oh, I get it. I will continue to keep on.

**Comment by Kumo-YZX** [comment]
@owocki Is current code eligible enough to be merged? Or I need to finish cordon-off-unnecessary-app work first?

**Comment by Kumo-YZX** [comment]
> @Kumo-YZX can you do the final cleanups if you are happy with this? Also, have you tested the existing functionality (tips, grants, bounties, etc.) and interactions between Twitter and Github users?

Sorry for the delay. There was something wrong when I tried to create test kudos. I will continue to keep on this.

**Comment by Kumo-YZX** [comment]
I have banned Twitter users from using GitHub functions.

**Comment by Kumo-YZX** [comment]
And the http://tokenfactory.surge.sh website does not work... at all... so, it is hard for me to test some functions... 
image

**Comment by Kumo-YZX** [comment]
So... despite this, there aren't other problems...

**Comment by owocki** [comment]
Kumo -- Thanks for all your hard work on this.  I'm going to pay this out now, hopefully well get it merged sometime soon.

**Comment by Kumo-YZX** [comment]
We need to be careful that twitter oauth api sometimes may fail and throw a 401 error, at that situation we need to reload the program. GitHub function is blocked for twitter users, and if there is a page without the block function, some message, such as "user not exist" will be returned due to the prefix "twittere_"..

**Comment by Kumo-YZX** [comment]
image
I am trying to apply for some Ether and continue on this..

**Comment by Kumo-YZX** [comment]
@lychees has submitted the work on gitcoin.co :)

**Comment by owocki** [comment]
just paid!

**Comment by Kumo-YZX** [comment]
Thanks. :)

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by thelostone-mc** [comment]
Closing this out for now as we don't mean to get it in!
Will get this reopened if we decide to set this up

**Pull request closed without merging**